distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso

```diff
@@ -310,68 +317,112 @@ type SQLServer_Dialect
-        base_expr = expr_null_checks_pair.first
-        built_expr = case expr of
-            SQL_Expression.Operation _ _ _ ->
-                null_check = if expr_null_checks_pair.second.length == 0 then "" else
-                    SQL_Builder.code "WHEN " ++ _generate_null_check_sql_builder expr_null_checks_pair.second ++ " THEN NULL "
-                SQL_Builder.code "CASE " ++ null_check ++ "WHEN " ++ base_expr ++ " THEN 1 ELSE 0 END"
-            _ -> base_expr
-        built_expr ++ Base_Generator.alias self name
+    generate_expression self base_gen expr:(SQL_Expression | Order_Descriptor | Query) for_select:Boolean -> SQL_Builder =
+        if for_select then _generate_column self base_gen expr else
+            _generate_expression self base_gen expr generate_as_bool=True . first
+
+## Returns a pair of a SQL_Builder for the given expression and a vector of columns
+   that have been used in the expression and need to be checked for nulls.
+   SQL Server needs special handling commpared to ther databases as it does not have a
+   boolean data type.
+   This means that you can write
+       SELECT * FROM MyTable WHERE [Column1] > [Column2]
+   but you cannot write
+       SELECT [Column1] > [Column2] FROM MyTable
+   to write the second query you need to write
+       SELECT CASE WHEN [Column1] IS NULL OR [Column2] IS NULL WHEN [Column1] > [Column2] THEN 1 ELSE 0 END FROM MyTable
+   The below function collects all of the fields which are needed to be checked for nulls returning them in a vector
+   as the second element of the pair.
+   The first element of the pair is the SQL_Builder for the expression.
+private _generate_expression dialect base_gen expr generate_as_bool = case expr of
+    SQL_Expression.Column _ _ ->
+        wrapped_name = base_gen.generate_expression dialect expr
+        typed_result = case generate_as_bool of 
+            False -> wrapped_name
+            True -> wrapped_name ++ " = 1"
+        pair typed_result [wrapped_name]
+    SQL_Expression.Constant value -> 
+        wrapped = case value of
+            Nothing -> SQL_Builder.code "NULL"
+            _ -> SQL_Builder.interpolation value
+        pair wrapped [wrapped]
+    SQL_Expression.Literal value -> 
+        wrapped = SQL_Builder.code value
+        pair wrapped [wrapped]
+    SQL_Expression.Text_Literal _ ->
+        wrapped_literal = base_gen.generate_expression dialect expr
+        pair wrapped_literal []
+    SQL_Expression.Operation kind arguments metadata ->
+        op = dialect.dialect_operations.operations_dict.get kind (Error.throw <| Unsupported_Database_Operation.Error kind)
+        requires_bool = _op_requires_bool kind
+        parsed_args_and_null_checks = arguments.map (c -> _generate_expression dialect base_gen c requires_bool)
+        parsed_args = parsed_args_and_null_checks.map .first
+
+        null_checks = parsed_args_and_null_checks.map .second  . flatten
```

Comment:
Maybe
```suggestion
        null_checks = parsed_args_and_null_checks.flat_map .second
```

both are equivalent here, I'm not arguing one is better than the other.